mithril-aggregator/src/services/epoch_service.rs

```diff
@@ -0,0 +1,847 @@
+            update_protocol_parameters_error: false,
+            precompute_epoch_data_error: false,
+        }
+    }
+
+    pub fn toggle_errors(
+        &mut self,
+        inform_epoch: bool,
+        update_protocol_parameters: bool,
+        precompute_epoch: bool,
+    ) {
+        self.inform_epoch_error = inform_epoch;
+        self.update_protocol_parameters_error = update_protocol_parameters;
+        self.precompute_epoch_data_error = precompute_epoch;
+    }
+
+    fn unwrap_data(&self) -> Result<&EpochData, EpochServiceError> {
+        self.epoch_data
+            .as_ref()
+            .ok_or(EpochServiceError::NotYetInitialized)
+    }
+
+    fn unwrap_computed_data(&self) -> Result<&ComputedEpochData, EpochServiceError> {
+        let epoch = self.unwrap_data()?.epoch;
+
+        self.computed_epoch_data
+            .as_ref()
+            .ok_or(EpochServiceError::NotYetComputed(epoch))
+    }
+}
+
+#[cfg(test)]
+#[async_trait]
+impl EpochService for FakeEpochService {
+    async fn inform_epoch(&mut self, epoch: Epoch) -> StdResult<()> {
+        if self.inform_epoch_error {
+            anyhow::bail!("inform_epoch fake error, given epoch: {epoch}");
+        }
+        Ok(())
+    }
+
+    async fn update_protocol_parameters(&mut self) -> StdResult<()> {
+        if self.update_protocol_parameters_error {
+            anyhow::bail!("update_protocol_parameters fake error");
+        }
+        Ok(())
+    }
+
+    async fn precompute_epoch_data(&mut self) -> StdResult<()> {
+        if self.precompute_epoch_data_error {
+            anyhow::bail!("precompute_epoch_data fake error");
+        }
+        Ok(())
+    }
+
+    fn epoch_of_current_data(&self) -> StdResult<Epoch> {
+        Ok(self.unwrap_data()?.epoch)
+    }
+
+    fn current_protocol_parameters(&self) -> StdResult<&ProtocolParameters> {
+        Ok(&self.unwrap_data()?.protocol_parameters)
+    }
+
+    fn next_protocol_parameters(&self) -> StdResult<&ProtocolParameters> {
+        Ok(&self.unwrap_data()?.next_protocol_parameters)
+    }
+
+    fn upcoming_protocol_parameters(&self) -> StdResult<&ProtocolParameters> {
+        Ok(&self.unwrap_data()?.upcoming_protocol_parameters)
+    }
+
+    fn current_aggregate_verification_key(&self) -> StdResult<&ProtocolAggregateVerificationKey> {
+        Ok(&self.unwrap_computed_data()?.aggregate_verification_key)
+    }
+
+    fn next_aggregate_verification_key(&self) -> StdResult<&ProtocolAggregateVerificationKey> {
+        Ok(&self.unwrap_computed_data()?.next_aggregate_verification_key)
+    }
+
+    fn current_signers_with_stake(&self) -> StdResult<&Vec<SignerWithStake>> {
+        Ok(&self.unwrap_data()?.signers)
+    }
+
+    fn next_signers_with_stake(&self) -> StdResult<&Vec<SignerWithStake>> {
+        Ok(&self.unwrap_data()?.next_signers)
+    }
+
+    fn protocol_multi_signer(&self) -> StdResult<&ProtocolMultiSigner> {
+        Ok(&self.unwrap_computed_data()?.protocol_multi_signer)
+    }
```

Comment:
Maybe you could have made a blanket implementation for these?

Reply:
I don't think that we want that, the trait should not define the `unwrap_xxx` methods since the unwrapped types are purposely privates making a blanket implementation impossible.